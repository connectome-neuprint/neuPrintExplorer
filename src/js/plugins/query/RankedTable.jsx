/* eslint-disable camelcase */
/*
 * Implements table view that shows ordered strongest connection to each neuron
 * and visually indicates the different classes of neurons.  (This is meant
 * to be similar to Lou's tables.)
 */
import React from 'react';
import PropTypes from 'prop-types';
import randomColor from 'randomcolor';

import Button from '@mui/material/Button';
import { withStyles } from '@material-ui/core/styles';
import Radio from '@mui/material/Radio';
import RadioGroup from '@mui/material/RadioGroup';
import FormControl from '@mui/material/FormControl';
import FormControlLabel from '@mui/material/FormControlLabel';
import Typography from '@mui/material/Typography';
import Icon from '@mui/material/Icon';
import Switch from '@mui/material/Switch';

import ColorBox from 'plugins/ColorBox';
import NeuronInputField from './shared/NeuronInputField';
import AdvancedNeuronInput from './shared/AdvancedNeuronInput';
import HPWeightSlider from './shared/HPWeightSlider';

const squareSize = 100;

const styles = () => ({
  textField: {},
  formControl: {
    margin: '0.5em 0 1em 0',
    width: '100%'
  }
});

const pluginName = 'RankedTable';
const pluginAbbrev = 'rt';

export class RankedTable extends React.Component {
  static get details() {
    return {
      name: pluginName,
      displayName: 'Ranked Table',
      abbr: pluginAbbrev,
      description: 'Show connections to neuron(s) ranked in order and colored by neuron class',
      visType: 'HeatMapTable'
    };
  }

  static fetchParameters() {
    return {
      queryString: '/npexplorer/rankedtable'
    };
  }

  static processResults({ query, apiResponse }) {
    const { dataSet } = query;
    const { useHighConfidence, find_inputs } = query.pm;

    const colorMap = {};
    const reverseCounts = {};

    const data = [];
    let lastBody = -1;
    let columns = [];
    let maxColumns = 0;

    // need to re-sort by weightHP (grouped by body1 id) if using high confidence
    if (useHighConfidence) {
      apiResponse.data.sort((a, b) => {
        const weightHPa = a[10];
        const weightHPb = b[10];
        const body1a = a[9];
        const body1b = b[9];

        if (body1a > body1b) {
          return 1;
        }
        if (body1a < body1b) {
          return -1;
        }
        if (body1a === body1b) {
          if (weightHPa > weightHPb) {
            return -1;
          }
          if (weightHPa < weightHPb) {
            return 1;
          }
        }
        return 0;
      });
    }

    apiResponse.data.forEach(row => {
      const [, , , , weight, body2, mId, nId, preId, body1, weightHP] = row;

      if (
        (find_inputs === false && (preId !== mId || nId === mId)) ||
        (find_inputs === true && preId === mId)
      ) {
        if (body2 in reverseCounts) {
          reverseCounts[String(body2)][String(body1)] = useHighConfidence ? weightHP : weight;
        } else {
          reverseCounts[String(body2)] = {};
          reverseCounts[String(body2)][String(body1)] = useHighConfidence ? weightHP : weight;
        }
      }
    });

    apiResponse.data.forEach((row, index) => {
      const [
        neuron1,
        ,
        neuron2,
        neuron2Type,
        weight,
        body2,
        mId,
        nId,
        preId,
        body1,
        weightHP
      ] = row;

      if (
        (find_inputs === false && preId === mId) ||
        (find_inputs === true && (preId !== mId || nId === mId))
      ) {
        // check the colormap for the current type.
        // if present use the existing color.
        // else add a new one from the list. Unless they have all been used,
        // then add a random one.
        if (neuron2Type && !(neuron2Type in colorMap)) {
          colorMap[neuron2Type] = randomColor({ luminosity: 'light', hue: 'random' });
        } else if (neuron2 && !(neuron2 in colorMap)) {
          colorMap[neuron2] = randomColor({ luminosity: 'light', hue: 'random' });
        }
        // color should be white if we don't have a type defined.
        let weightColor = '#ffffff';
        if (neuron2Type) {
          weightColor = colorMap[neuron2Type];
        } else if (neuron2) {
          weightColor = colorMap[neuron2];
        }

        // start a new row for each body1.
        if (body1 !== lastBody) {
          data.push(columns);
          maxColumns = Math.max(columns.length - 1, maxColumns);
          columns = [`${neuron1 || body1} (${body1})`];
        }

        let reverseWeight = 0;
        if (body2 in reverseCounts && body1 in reverseCounts[body2]) {
          reverseWeight = reverseCounts[body2][body1];
        }

        const cellKey = `${index}${body1}${body2}`;

        // set arrow icons
        const arrow1 = find_inputs ? (
          <Icon fontSize="inherit">arrow_downward</Icon>
        ) : (
          <Icon fontSize="inherit">arrow_upward</Icon>
        );
        const arrow2 = find_inputs ? (
          <Icon fontSize="inherit">arrow_upward</Icon>
        ) : (
          <Icon fontSize="inherit">arrow_downward</Icon>
        );

        // truncate name if necessary
        let displayName = neuron2 || body2;
        if (displayName && displayName.length > 20) {
          displayName = `${displayName.substring(0, 17)}...`;
        }

        // add the current item to the current columns array
        columns.push({
          value: (
            <ColorBox
              margin={0}
              width={squareSize}
              height={squareSize}
              backgroundColor={weightColor}
              title=""
              key={cellKey}
              text={
                <>
                  <Typography
                    style={{
                      marginTop: '20px',
                      marginLeft: '10px',
                      marginRight: '10px',
                      textAlign: 'center',
                      wordBreak: 'break-all'
                    }}
                    variant="body2"
                    title={neuron2 || body2}
                  >
                    {displayName}
                  </Typography>
                  <Typography
                    style={{
                      marginTop: '0px',
                      marginLeft: '10px',
                      marginRight: '10px',
                      textAlign: 'center',
                      wordBreak: 'break-all'
                    }}
                    variant="caption"
                  >
                    {body2}
                  </Typography>

                  <div
                    style={{
                      marginTop: 'auto',
                      display: 'flex'
                    }}
                  >
                    <Typography
                      style={{
                        display: 'flex',
                        alignItems: 'center'
                      }}
                      variant="caption"
                    >
                      {arrow1}
                      {useHighConfidence ? weightHP : weight}
                    </Typography>
                    <Typography
                      style={{
                        display: 'flex',
                        alignItems: 'center'
                      }}
                      variant="caption"
                    >
                      {arrow2}
                      {reverseWeight}
                    </Typography>
                  </div>
                </>
              }
            />
          ),
          uniqueId: cellKey
        });

        lastBody = body1;
      }
    });
    // make sure the last column is added.
    maxColumns = Math.max(columns.length - 1, maxColumns);
    data.push(columns);

    const headings = Array(maxColumns)
      .fill(0)
      .map((e, i) => `#${i + 1}`);

    const neuronSrc = query.pm.neuron_name || query.pm.neuron_id;
    let title = `Neurons postsynaptic to ${neuronSrc}`;

    if (query.pm.find_inputs) {
      title = `Neurons presynaptic to ${neuronSrc}`;
    }

    return {
      columns: ['', ...headings],
      data,
      debug: apiResponse.debug,
      dataSet,
      title
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      neuronSrc: '',
      preOrPost: 'pre',
      useHighConfidence: false,
      advancedSearch: false
    };
  }

  processRequest = () => {
    const { neuronSrc, preOrPost, useHighConfidence, advancedSearch } = this.state;
    const { dataSet, actions, submit } = this.props;
    if (neuronSrc !== '') {
      const parameters = { dataset: dataSet };
      if (/^\d+$/.test(neuronSrc)) {
        parameters.neuron_id = neuronSrc;
      } else {
        parameters.neuron_name = neuronSrc;
      }

      if (!advancedSearch) {
        parameters.enable_contains = true;
      }

      if (preOrPost === 'pre') {
        parameters.find_inputs = false;
      } else {
        parameters.find_inputs = true;
      }

      parameters.useHighConfidence = useHighConfidence;

      const query = {
        dataSet,
        visProps: { squareSize },
        plugin: pluginName,
        pluginCode: pluginAbbrev,
        parameters
      };

      submit(query);
    } else {
      actions.formError('Please enter a neuron name.');
    }
  };

  addNeuron = neuronSrc => {
    this.setState({ neuronSrc });
  };

  setDirection = event => {
    this.setState({ preOrPost: event.target.value });
  };

  toggleHighConfidence = () => {
    const { useHighConfidence } = this.state;
    this.setState({ useHighConfidence: !useHighConfidence });
  };

  toggleAdvanced = event => {
    this.setState({ advancedSearch: event.target.checked, neuronSrc: '' });
  };

  render() {
    const { classes, isQuerying, isPublic, dataSet } = this.props;
    const { neuronSrc, preOrPost, useHighConfidence, advancedSearch } = this.state;
    return (
      <div>
        {advancedSearch ? (
          <AdvancedNeuronInput
            onChange={this.addNeuron}
            value={neuronSrc}
            dataSet={dataSet}
            handleSubmit={this.processRequest}
          />
        ) : (
          <NeuronInputField
            onChange={this.addNeuron}
            value={neuronSrc}
            dataSet={dataSet}
            handleSubmit={this.processRequest}
          />
        )}
        <FormControl className={classes.formControl}>
          <FormControlLabel
            control={
              <Switch checked={advancedSearch} onChange={this.toggleAdvanced} color="primary" />
            }
            label={
              <Typography variant="subtitle1" style={{ display: 'inline-flex' }}>
                Advanced input
              </Typography>
            }
          />
        </FormControl>
        <FormControl component="fieldset" required className={classes.formControl}>
          <RadioGroup
            aria-label="preorpost"
            name="preorpost"
            className={classes.group}
            value={preOrPost}
            onChange={this.setDirection}
          >
            <FormControlLabel
              value="pre"
              control={<Radio color="primary" />}
              label="Rank outputs"
            />
            <FormControlLabel
              value="post"
              control={<Radio color="primary" />}
              label="Rank inputs"
            />
          </RadioGroup>
        </FormControl>
        {isPublic ? (
          ''
        ) : (
          <HPWeightSlider
            formControlClass={classes.formControl}
            useHighConfidence={useHighConfidence}
            toggleHighConfidence={this.toggleHighConfidence}
          />
        )}
        <Button
          disabled={isQuerying}
          variant="contained"
          color="primary"
          onClick={this.processRequest}
        >
          Submit
        </Button>
      </div>
    );
  }
}

RankedTable.propTypes = {
  isQuerying: PropTypes.bool.isRequired,
  submit: PropTypes.func.isRequired,
  actions: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  dataSet: PropTypes.string.isRequired,
  isPublic: PropTypes.bool.isRequired
};

export default withStyles(styles)(RankedTable);
