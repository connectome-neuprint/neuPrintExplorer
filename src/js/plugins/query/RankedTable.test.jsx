import React from 'react'; // React import needed for JSX
import { render } from '../../tests/test-utils';
import { RankedTable } from './RankedTable';

const styles = { formControl: 'formControlClass' };
const { actions, submit } = global;

const neoServerSettings = {
  get: () => 'http://example.com'
};

const component = (
  <RankedTable
    availableROIs={['roiA', 'roiB', 'roiC']}
    dataSet="test"
    datasetstr="test"
    actions={actions}
    submit={submit}
    classes={styles}
    history={{ push: jest.fn() }}
    isQuerying={false}
    neoServerSettings={neoServerSettings}
    neoServer="testServer"
    isPublic={false}
  />
);

const componentPublic = (
  <RankedTable
    availableROIs={['roiA', 'roiB', 'roiC']}
    dataSet="test"
    datasetstr="test"
    actions={actions}
    submit={submit}
    classes={styles}
    history={{ push: jest.fn() }}
    isQuerying={false}
    neoServerSettings={neoServerSettings}
    neoServer="testServer"
    isPublic
  />
);

describe('RankedTable Plugin', () => {
  it('should have required details fields', () => {
    expect(RankedTable.details.name).toBeTruthy();
    expect(RankedTable.details.description).toBeTruthy();
  });
  it('renders correctly', () => {
    const { asFragment} = render(component);
    expect(asFragment()).toMatchSnapshot(); // Snapshot test
  });
  it('renders correctly in public mode', () => {
    const { asFragment} = render(componentPublic);
    expect(asFragment()).toMatchSnapshot(); // Snapshot test
  });
});
