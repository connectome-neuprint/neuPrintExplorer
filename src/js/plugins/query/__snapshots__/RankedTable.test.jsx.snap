// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RankedTable Plugin renders correctly 1`] = `
<DocumentFragment>
  <div>
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-1tkmop-MuiFormLabel-root-MuiInputLabel-root"
      for="select-multiple-chip"
    >
      Neuron Instance, Type or BodyId (optional)
    </label>
    <div
      class="NeuronInputField-select-1 css-b62m3t-container"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-2-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        class="asyncSelect__control css-13cymwt-control"
      >
        <div
          class="asyncSelect__value-container css-hlgwow"
        >
          <div
            class="asyncSelect__placeholder css-1jqq78o-placeholder"
            id="react-select-2-placeholder"
          >
            Type or Paste text for options
          </div>
          <div
            class="asyncSelect__input-container css-19bb58m"
            data-value=""
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-describedby="react-select-2-placeholder"
              aria-expanded="false"
              aria-haspopup="true"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              class="asyncSelect__input"
              id="react-select-2-input"
              role="combobox"
              spellcheck="false"
              style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="asyncSelect__indicators css-1wy0on6"
        >
          <span
            class="asyncSelect__indicator-separator css-1u9des2-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            class="asyncSelect__indicator asyncSelect__dropdown-indicator css-1xc3v61-indicatorContainer"
          >
            <svg
              aria-hidden="true"
              class="css-tj5bde-Svg"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root formControlClass css-1iw3t7y-MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-rehya9-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-161ms7l-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              class="PrivateSwitchBase-input MuiSwitch-input css-j8yymo"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-17jyosd-MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiSwitch-track css-tj8d99-MuiSwitch-track"
          />
        </span>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
          style="display: inline-flex;"
        >
          Advanced input
        </h6>
      </label>
    </div>
    <fieldset
      class="MuiFormControl-root formControlClass css-1iw3t7y-MuiFormControl-root"
    >
      <div
        aria-label="preorpost"
        class="MuiFormGroup-root MuiRadioGroup-root css-1mac7pt-MuiFormGroup-root"
        role="radiogroup"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
        >
          <span
            class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-13lvt8g-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-j8yymo"
              name="preorpost"
              type="radio"
              value="pre"
            />
            <span
              class="css-1qiat4j"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1rr0-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-z8nmqa-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
          >
            Rank outputs
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
        >
          <span
            class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-13lvt8g-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              class="PrivateSwitchBase-input css-j8yymo"
              name="preorpost"
              type="radio"
              value="post"
            />
            <span
              class="css-1qiat4j"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1rr0-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-o8rdeq-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
          >
            Rank inputs
          </span>
        </label>
      </div>
    </fieldset>
    <div
      class="MuiFormControl-root formControlClass css-1iw3t7y-MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-rehya9-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-161ms7l-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              class="PrivateSwitchBase-input MuiSwitch-input css-j8yymo"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-17jyosd-MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiSwitch-track css-tj8d99-MuiSwitch-track"
          />
        </span>
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          style="display: inline-flex;"
        >
          Limit to high-confidence synapses
        </p>
      </label>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;

exports[`RankedTable Plugin renders correctly in public mode 1`] = `
<DocumentFragment>
  <div>
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-1tkmop-MuiFormLabel-root-MuiInputLabel-root"
      for="select-multiple-chip"
    >
      Neuron Instance, Type or BodyId (optional)
    </label>
    <div
      class="NeuronInputField-select-2 css-b62m3t-container"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-3-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        class="asyncSelect__control css-13cymwt-control"
      >
        <div
          class="asyncSelect__value-container css-hlgwow"
        >
          <div
            class="asyncSelect__placeholder css-1jqq78o-placeholder"
            id="react-select-3-placeholder"
          >
            Type or Paste text for options
          </div>
          <div
            class="asyncSelect__input-container css-19bb58m"
            data-value=""
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-describedby="react-select-3-placeholder"
              aria-expanded="false"
              aria-haspopup="true"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              class="asyncSelect__input"
              id="react-select-3-input"
              role="combobox"
              spellcheck="false"
              style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="asyncSelect__indicators css-1wy0on6"
        >
          <span
            class="asyncSelect__indicator-separator css-1u9des2-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            class="asyncSelect__indicator asyncSelect__dropdown-indicator css-1xc3v61-indicatorContainer"
          >
            <svg
              aria-hidden="true"
              class="css-tj5bde-Svg"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root formControlClass css-1iw3t7y-MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-rehya9-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-161ms7l-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              class="PrivateSwitchBase-input MuiSwitch-input css-j8yymo"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-17jyosd-MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiSwitch-track css-tj8d99-MuiSwitch-track"
          />
        </span>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
          style="display: inline-flex;"
        >
          Advanced input
        </h6>
      </label>
    </div>
    <fieldset
      class="MuiFormControl-root formControlClass css-1iw3t7y-MuiFormControl-root"
    >
      <div
        aria-label="preorpost"
        class="MuiFormGroup-root MuiRadioGroup-root css-1mac7pt-MuiFormGroup-root"
        role="radiogroup"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
        >
          <span
            class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-13lvt8g-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-j8yymo"
              name="preorpost"
              type="radio"
              value="pre"
            />
            <span
              class="css-1qiat4j"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1rr0-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-z8nmqa-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
          >
            Rank outputs
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
        >
          <span
            class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-13lvt8g-MuiButtonBase-root-MuiRadio-root"
          >
            <input
              class="PrivateSwitchBase-input css-j8yymo"
              name="preorpost"
              type="radio"
              value="post"
            />
            <span
              class="css-1qiat4j"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1rr0-MuiSvgIcon-root"
                data-testid="RadioButtonUncheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-o8rdeq-MuiSvgIcon-root"
                data-testid="RadioButtonCheckedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
          >
            Rank inputs
          </span>
        </label>
      </div>
    </fieldset>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
